[forward ad] Sync conj between primal and tangent on set forward grad

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78358

Fixes https://github.com/pytorch/pytorch/issues/78256

New changes:
- I realized that `copy_` basically does most of what we want: the conj/neg bit is preserved, and the resulting value is correct
- Fixed the forward AD formula for `_neg_view` previously the tangent did not perform a view
- `_new_zeros_with_same_feature_meta` should be a fallthrough for neg/conj keys because we want to be able to see the underlying storage when we create the new zeros
- Removed the `.resolve_conj` from the previous iteration because we're creating a new-zeros anyway so the values/conj bit does not matter

soulitzer committed:
[forward ad] Sync conj for between primal and tangent on set forward grad

[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -6619,6 +6619,14 @@ def test_metadata_check_when_primal_has_neg_bit(self):
             dual[1:]
 
+    def test_metadata_check_check_conj(self):
+        x = torch.randn(2, 3, 4, dtype=torch.cdouble, requires_grad=True)
```
TODO: requires_grad not necessary

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


TODO: test view over in-place case?

[ghstack-poisoned]

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


TODO: test view over in-place case?

[ghstack-poisoned]

Lezcano commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -97,9 +97,23 @@ namespace {
     if (!at::_has_same_storage_numel(base, other)) {
       return false;
     }
+    if (base.is_conj() != other.is_conj()) {
+      return false;
+    }
```
We'll need to do the same thing with `is_neg`, otherwise if you do `at::imag(a.conj())` we're going to run into similar errors I reckon.

lezcano reviewed (approved):
The PR looks good. Would it be possible to add a test that fails on the case we were not testing before?
Also, I realised that we need to do the same thing for the negative bit (view onto the imaginary part of a conjugated tensor). Could you add this either to this PR or to another PR of the stack?

soulitzer commented:
Yup, the check for the case that wasn't covered has been added. And yeah we should definitely mirror this change for the neg case.

albanD reviewed (commented):
Agreed that the same should be done for neg!

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


TODO: test view over in-place case?

Fixes https://github.com/pytorch/pytorch/issues/78256


[ghstack-poisoned]

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


TODO: test view over in-place case?

Fixes https://github.com/pytorch/pytorch/issues/78256


[ghstack-poisoned]

Lezcano commented on test/test_autograd.py:
```diff
@@ -6613,6 +6613,28 @@ def test_metadata_check_when_primal_has_neg_bit(self):
+                with fwAD.dual_level():
+                    dual = fwAD.make_dual(x, t)
+                    self.assertTrue(fwAD.unpack_dual(dual).tangent is not t)
+                    torch.real(dual)
```
Add a `torch.imag(dual)` as well here and in the other branch, please.

Lezcano commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -97,9 +97,40 @@ namespace {
+  Tensor maybe_match_conj_or_neg(const Tensor& other, const Tensor& base) {
+    if (base.is_conj()) {
+      if (other.is_neg()) {
+        return other.resolve_neg().conj_physical().conj();
```
Doesn't `conj_physical` do the `resolve_neg` as well? @anjali411

lezcano reviewed (commented):
A small comment and a small performance question, otherwise this looks good

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


TODO: test view over in-place case?

Fixes https://github.com/pytorch/pytorch/issues/78256


[ghstack-poisoned]

soulitzer commented:
See updated PR description for latest changes

lezcano reviewed (approved):
Wow, great golfing skills! Thank you!

soulitzer commented:
Running into an issue with batched tensors (in core): when `conj` is performed on a batched tensor, only the conj key of the inner tensor is updated. `vmap(jvp` in functorch still seems to work fine though for the failing ops. Not sure how or whether we should try to fix this yet.

anjali411 commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -97,9 +97,22 @@ namespace {
   }
 
+  Tensor maybe_match_conj_or_neg(const Tensor& self, const Tensor& target) {
+    if (target.is_conj()) {
+      return self.conj();
+    } else if (target.is_neg()) {
```
It's because no current public API allows you to create a tensor that has both conj and neg bit set

soulitzer commented:
cc @zou3519

zou3519 commented:
You can skip the tests for BatchedTensors in core, but I want to be really sure that all of the functorch jvp tests still pass after this

soulitzer commented:
Repro:

```
import torch
import torch.autograd.forward_ad as fwAD
from torch._vmap_internals import _vmap

# Succeeds (no vmap)

primal = torch.rand(1, 6, dtype=torch.cfloat).conj()
tangent = torch.rand(1, 6, dtype=torch.cfloat)

with fwAD.dual_level():
    dual = fwAD.make_dual(primal, tangent)
    print("primal: ", dual.is_conj(), "tangent: ", fwAD.unpack_dual(dual).tangent.is_conj())
    dual[1:]

# Fails (with vmap)
primal = torch.rand(1, 6, dtype=torch.cfloat).conj()
tangents = torch.rand(16, 1, 6, dtype=torch.cfloat)

def jvp(tangent):
    with fwAD.dual_level():
        dual = fwAD.make_dual(primal, tangent)
        print("primal: ", dual.is_conj(), "tangent: ", fwAD.unpack_dual(dual).tangent.is_conj())
        dual[1:]
        return fwAD.unpack_dual(dual).tangent

_vmap(jvp, 0, 0)(tangents)

```
The second part prints `primal: True tangent: False` and errors with:
`RuntimeError: !this_view_meta->has_fw_view() INTERNAL ASSERT FAILED at "../torch/csrc/autograd/autograd_meta.cpp":218, please report a bug to PyTorch. Expected the output of forward differentiable view operations to have the tangent have the same layout as primal`

We expect the second part to also print `primal: True tangent: True` because `make_dual` is supposed to match the tangent's conj-ness with that of the primal.

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


Fixes https://github.com/pytorch/pytorch/issues/78256

New changes:
- I realized that `copy_` basically does most of what we want: the conj/neg bit is preserved, and the resulting value is correct
- Fixed the forward AD formula for `_neg_view` previously the tangent did not perform a view
- `_new_zeros_with_same_feature_meta` should be a fallthrough for neg/conj keys because we want to be able to see the underlying storage when we create the new zeros
- Removed the `.resolve_conj` from the previous iteration because we're creating a new-zeros anyway so the values/conj bit does not matter




[ghstack-poisoned]

zou3519 reviewed (approved):
lgtm from the functorch side, thanks for testing it out, Jeffrey

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "[forward ad] Sync conj between primal and tangent on set forward grad"


Fixes https://github.com/pytorch/pytorch/issues/78256

New changes:
- I realized that `copy_` basically does most of what we want: the conj/neg bit is preserved, and the resulting value is correct
- Fixed the forward AD formula for `_neg_view` previously the tangent did not perform a view
- `_new_zeros_with_same_feature_meta` should be a fallthrough for neg/conj keys because we want to be able to see the underlying storage when we create the new zeros
- Removed the `.resolve_conj` from the previous iteration because we're creating a new-zeros anyway so the values/conj bit does not matter




[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

anjali411 commented on test/test_autograd.py:
```diff
@@ -6621,6 +6621,30 @@ def test_metadata_check_when_primal_has_neg_bit(self):
+        keys = {
+            "NEITHER": lambda x: x,
+            "CONJ": lambda x: x.conj(),
+            "NEG": lambda x: x._neg_view()
```
The way to get neg bit in the public API is: x.conj().imag

soulitzer commented on torch/csrc/autograd/autograd_meta.cpp:
```diff
@@ -163,6 +176,7 @@ void AutogradMeta::set_fw_grad(const at::TensorBase& new_grad_base, const at::Te
             new_base_fw_grad = new_grad;
           } else {
             new_base_fw_grad = at::_new_zeros_with_same_feature_meta(new_grad, base);
+            new_base_fw_grad = maybe_match_conj_or_neg(new_base_fw_grad, base);
```
Oh cool, thanks!
